plugin/cog/cog-streams/src/main/java/it/geosolutions/imageioimpl/plugins/cog/DefaultCogImageInputStream.java

```diff
@@ -0,0 +1,399 @@
+/*
+ *    ImageI/O-Ext - OpenSource Java Image translation Library
+ *    http://www.geo-solutions.it/
+ *    http://java.net/projects/imageio-ext/
+ *    (C) 2019, GeoSolutions
+ *
+ *    This library is free software; you can redistribute it and/or
+ *    modify it under the terms of the GNU Lesser General Public
+ *    License as published by the Free Software Foundation;
+ *    either version 3 of the License, or (at your option) any later version.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package it.geosolutions.imageioimpl.plugins.cog;
+
+import it.geosolutions.imageio.plugins.cog.CogImageReadParam;
+
+import javax.imageio.stream.IIOByteBuffer;
+import javax.imageio.stream.ImageInputStream;
+import javax.imageio.stream.MemoryCacheImageInputStream;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URL;
+import java.nio.ByteOrder;
+import java.util.Set;
+import java.util.logging.Logger;
+
+import static it.geosolutions.imageioimpl.plugins.cog.CogTileInfo.HEADER_TILE_INDEX;
+
+/**
+ * ImageInputStream implementation for COG.  This class will asynchronously request all requested ranges be read by
+ * the provided RangeReader implementation and cache the results in a delegate MemoryCacheImageInputStream.  When
+ * TIFFImageReader requests tiles, all of the data will be already prefetched and ready to be used.
+ *
+ * @author joshfix
+ * Created on 2019-08-23
+ */
+public class DefaultCogImageInputStream implements ImageInputStream, CogImageInputStream {
```

Comment:
So this class works in two steps, initially it has no content and won't work, then one inits the header, and it can provide the header only, and after calling readranges, it can offer the tiles too (only the requested ones). So it's not a generic image input stream, but a special purpose one that can only be used in concert with the COG reader. This should be explicit imho.

Reply:
Also all correct.  The central concept of my design is to "warm" the streams with the data of (only) the requested tiles up front, before we actually start to decode tiles.  The only way that can be accomplished is by reading the header to determine the tile offsets and lengths.  Basically the input streams are completely non-functional without being initialized with the header data. 

I guess I assumed from the name and the interface it would be apparent that it was purpose-built for COG, but I will add more documentation stating and explaining its purpose and limitations.

Reply:
ImageInputStream are an extension point that can be looked up and used in whatever other format, making one that's specific is off its intended usage. I understand the reason behind the current implementation, and hopefully it's not going to cause side effects for other concurrent sources in the same server (e.g., other geotiffs, ecws and the like), but it's best to be quite clear when going off design.

Reply:
Added explicit documentation regarding purpose.